sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -84,26 +86,93 @@ static RT::PiProgram createSpirvProgram(const RT::PiContext Context,
   return Program;
 }
 
+template <typename RetT>
+RetT *
+waitUntilBuilt(KernelProgramCache &Cache,
+               KernelProgramCache::EntityWithState<RetT> *WithBuildState) {
+  // condition_variable.wait() call is wrapped with a while due to common
+  // cond_var for building all kernels and programs
+  int State = WithBuildState->State.load();
```

Comment:
Since you check for a particular WithBuildState object, you return from waitUntilBuilt only when the required program is built. The outer while loop seems to be unnecessary.

Reply:
The outer loop elimitnates unwanted locks of mutex for cv.wait() if the kernel/program is already built.
This method is only called if `Inserted` is equal to `false`.

Reply:
Another case is described in the comment prior to the loop.
Say, thread T1 is building kernel K1, thread T2 awaits kernel K2 is built, thread T3 builds kernel K2.
// Note: `condition_variable.wait()` methods are guaranteed to return on call to `condition_variable.notify_all()`.
If T1 finishes built process sooner than T3, T1 will call `cond_var.notify_all()` and T2 will return from `cond_var.wait()`.

Reply:
> If T1 finishes built process sooner than T3, T1 will call `cond_var.notify_all()` and T2 will return from `cond_var.wait()`.

Why? `State` of T1 and T2 are different, so T2 will be awaken, but it will not return until its condition (`State == BS_Done` || State == BS_Failed`) is true.

Reply:
Sure. My mistake.